### PR TITLE
3611: Add test for journey map event consistency

### PR DESCRIPTION
**Pull request by MikeCollingwood**
### What changed

- Test all CRI states have the same events handled
- Test all states for the same page have the same events handled
- Test events consistent entering and exiting nested journeys
- Update CRI states with Error targets for events not handled

### Why did it change

- To make interactions with CRIs more consistent
- To know nested journeys have events covered
- To catch inconsistent journey map behaviour

### Issue tracking

- [PYIC-3611](https://govukverify.atlassian.net/browse/PYIC-3611)

### Environment variables or secrets

- [ ] No environment variables or secrets were added or changed

### Other considerations

- [ ] Update [README](./blob/main/README.md) with any new instructions or tasks

[PYIC-3611]: https://govukverify.atlassian.net/browse/PYIC-3611?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+        List<StateAndEvents> criStates = new ArrayList<>();
+        Set<String> criStateEvents = new HashSet<>();
+        findCriStatesAndEvents(stateMachine, criStates, criStateEvents);
+
+        for (StateAndEvents stateAndEvents : criStates) {
+            assertEquals(
+                    criStateEvents,
+                    stateAndEvents.events,
+                    String.format("%s doesn't handle all CRI state events", stateAndEvents.state));
```
Comment (Joe-Edwards-GDS) [inline lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java]: Does the message describe which events are missing? Feels like that would be really useful if possible (but also not worth spending ages on)

(Same with the test below)

File: lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    void shouldMatchNestedJourneyEntryEvents(IpvJourneyTypes journeyType) throws IOException {
+        StateMachineInitializer stateMachineInitialiser = new StateMachineInitializer(journeyType);
+
+        Map<String, State> stateMachine = stateMachineInitialiser.initialize();
+        Set<String> stateMachineKeys = stateMachine.keySet();
+
+        for (String key : stateMachineKeys) {
+            State state = stateMachine.get(key);
+
+            if (state instanceof NestedJourneyInvokeState) {
```
Comment (Joe-Edwards-GDS) [inline lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java]: I think we might be able to use some more fancy Java features here to avoid the explicit cast:

```
if (state instanceof NestedJourneyInvokeState nestedState) {
  Set<String> entryEvents = nestedState.getExitEvents().keySet();
  ...
}
```
I'm not super familiar with the syntax though!

File: lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+                            if (event instanceof BasicEvent) {
+                                if (((BasicEvent) event).getTargetState() == key) {
+                                    assertTrue(entryEvents.contains(eventName));
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            ;
```
Comment (Joe-Edwards-GDS) [inline lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java]: Weird dangling semis?

File: lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+            State state = stateMachine.get(key);
+
+            if (state instanceof NestedJourneyInvokeState) {
+                Set<String> entryEvents =
+                        ((NestedJourneyInvokeState) state).getExitEvents().keySet();
+
+                for (String comparingStateKey : stateMachineKeys) {
+                    State comparingState = stateMachine.get(comparingStateKey);
+
+                    if (comparingState instanceof BasicState) {
```
Comment (Joe-Edwards-GDS) [inline lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java]: I think this does assume that we only ever link to nested states from basic states(?) - which is fine for now, but something to be aware of!

File: lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java (hunk 2 above)
Comment (Joe-Edwards-GDS) [reply lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java]: (same applies elsewhere)

File: lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    @ParameterizedTest
+    @EnumSource
+    void shouldMatchNestedJourneyEntryEvents(IpvJourneyTypes journeyType) throws IOException {
+        StateMachineInitializer stateMachineInitialiser = new StateMachineInitializer(journeyType);
+
+        Map<String, State> stateMachine = stateMachineInitialiser.initialize();
+        Set<String> stateMachineKeys = stateMachine.keySet();
+
+        for (String key : stateMachineKeys) {
+            State state = stateMachine.get(key);
```
Comment (Joe-Edwards-GDS) [inline lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java]: We could probably tighten up a few of these names to make it easier to understand what's going on here?

E.g. here we're looking for the `targetNestedState`, and further down we're looking for (any) `sourceBasicState` - maybe easier in combination with the pattern matching I describe below.

Comment (Joe-Edwards-GDS) [reply lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java]: (rationale being that we're verifying the events line up from 'source' to 'target')

File: lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+                    if (comparingState instanceof BasicState) {
+                        Map<String, Event> eventMap = ((BasicState) comparingState).getEvents();
+
+                        for (Map.Entry<String, Event> entry : eventMap.entrySet()) {
+                            String eventName = entry.getKey();
+                            Event event = entry.getValue();
+
+                            if (event instanceof BasicEvent) {
+                                if (((BasicEvent) event).getTargetState() == key) {
+                                    assertTrue(entryEvents.contains(eventName));
```
Comment (Joe-Edwards-GDS) [inline lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java]: (Does this give a useful failure message?)

File: lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+                        ((NestedJourneyInvokeState) state)
+                                .getNestedJourneyDefinition()
+                                .getNestedJourneyStates();
+                findPageSpecificStatesAndEvents(nestedStateMachine, pageStateMap);
+            }
+        }
+    }
+}
+
+class StateAndEvents {
```
Comment (Joe-Edwards-GDS) [inline lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java]: More modern Java, can we just make this a record class?

```
record StateAndEvents(String state, Set<String> events) {}
```

File: lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/ipv-core-main-journey.yaml [hunk 8]
```diff
@@ -137,6 +137,16 @@ CRI_F2F:
...
+    not-found:
+      targetState: ERROR
+    temporarily-unavailable:
+      targetState: ERROR
+    fail-with-no-ci:
+      targetState: ERROR
+    end:
+      targetState: ERROR
+    enhanced-verification:
+      targetState: ERROR
```
Comment (Joe-Edwards-GDS) [inline lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/ipv-core-main-journey.yaml]: Interestingly, once we add F2F CIMIT this should become `F2F_HANDOFF_PAGE` (because the F2F is a valid enhanced verification journey).

File: lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/ipv-core-main-journey.yaml [hunk 9]
```diff
@@ -137,6 +137,16 @@ CRI_F2F:
       targetState: F2F_HANDOFF_PAGE
     access-denied:
       targetState: PYI_ANOTHER_WAY
+    not-found:
+      targetState: ERROR
+    temporarily-unavailable:
```
Comment (Joe-Edwards-GDS) [inline lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/ipv-core-main-journey.yaml]: I'm curious how you found this - I think there is default behaviour here from the parent state which should be enough to pass the test(?)

If we're actively changing behaviour, I think we should probably separate that into a different PR - to keep this one as 'test only'.

Comment (Joe-Edwards-GDS) [reply lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/ipv-core-main-journey.yaml]: The 'end' case is particularly interesting - and we probably want to talk to some other people about whether an unexpected 'end' should just complete the journey, continue the journey as-is, or be an error (acknowledging that the user can probably just restart the journey and get a valid profile)

File: lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/nested-journey-definitions.yaml [hunk 10]
```diff
@@ -21,3 +33,13 @@ ADDRESS_AND_FRAUD:
...
+        not-found:
+          targetState: ERROR
+        temporarily-unavailable:
+          targetState: ERROR
+        fail-with-no-ci:
+          targetState: ERROR
+        enhanced-verification:
+          targetState: ERROR
+        access-denied:
+          targetState: ERROR
```
Comment (Wynndow) [inline lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/nested-journey-definitions.yaml]: Part of me is wondering if rather than having all this repetition on every state, we could define one parent state that has all events we know about with a target state of error. And then all other states inherit it and override just the events we expect to receive. Might make the expected behaviour for each state more obvious while still handling events that *could* be thrown.

Comment (Joe-Edwards-GDS) [reply lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/nested-journey-definitions.yaml]: Yeah - that could be a reasonable compromise. Having the repetition like this (and the test to validate) forces you to think about every state for every CRI, but perhaps it's unnecessarily tedious!

(It's still a lot better than having ad-hoc events all over the place)

Comment (Joe-Edwards-GDS) [reply lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/nested-journey-definitions.yaml]: This is probably where I'd be keen for @blakeyp 's opinion, since HILUX might be at the receiving end of most changes like this :)

File: lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java (hunk 1 above)
Comment (MikeCollingwood) [reply lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java]: The assertion displays expected and actual, so the tester can understand what's needed but I display the difference easily so will do!

File: lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java (hunk 2 above)
Comment (MikeCollingwood) [reply lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java]: I'm not familiar with the syntax either so I'll look into it because explicit casting didn't sit well - thank you!

File: lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java (hunk 4 above)
Comment (MikeCollingwood) [reply lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java:106]: Hmm, maybe I'll check it's not an InvokeNestedJourneyState instead! At least it'll break if a new state type comes in (I think all states are Basic/ Invoke nested ...)

File: lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java (hunk 6 above)
Comment (MikeCollingwood) [reply lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java]: Good point I need to add it in

File: lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/nested-journey-definitions.yaml (hunk 10 above)
Comment (blakeyp) [reply lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/nested-journey-definitions.yaml]: I like @Wynndow's suggestion of having the defaults on a parent state and each state just overrides any it needs to, will make it more readable and is consistent with how we've been using `CRI_STATE` as a parent

File: lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/ipv-core-main-journey.yaml (hunk 9 above)
Comment (MikeCollingwood) [reply lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/ipv-core-main-journey.yaml]: In the case where we're using the parent CRI_STATE to set the target state of the end event, I think it should be ERROR, since we don't have a consistent action on end

File: lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/nested-journey-definitions.yaml (hunk 10 above)
Comment (MikeCollingwood) [reply lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/nested-journey-definitions.yaml]: I've moved additional events into the parent: CRI_STATE, to complete the set

File: lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/ipv-core-main-journey.yaml (hunk 9 above)
Comment (MikeCollingwood) [reply lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/ipv-core-main-journey.yaml]: Could it expected where it relies on the default of going to IPV_SUCCESS_PAGE?

File: lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/ipv-core-main-journey.yaml [hunk 11]
```diff
@@ -6,10 +6,18 @@ END_JOURNEY:
...
+    not-found:
+      targetState: ERROR
+    fail-with-no-ci:
+      targetState: ERROR
     error:
       targetState: ERROR
     access-denied:
       targetState: PYI_NO_MATCH
+    enhanced-verification:
+      targetState: ERROR
```
Comment (Joe-Edwards-GDS) [inline lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/ipv-core-main-journey.yaml:20]: I think we might want to change some of these defaults - e.g. so the failure cases route to `PYI_NO_MATCH` or `PYI_ANOTHER_WAY`, but in the spirit of separating the tests from functional changes, I'll raise a new ticket.

File: lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/ipv-core-main-journey.yaml (hunk 9 above)
Comment (MikeCollingwood) [reply lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/ipv-core-main-journey.yaml]: Having talked to Patrick, since it shouldn't be reached we probably want to indicate this as an error so we can react to it. I'll update it to ERROR instead of the default IPV_SUCCESS_PAGE

**Review by Joe-Edwards-GDS — APPROVED** [review]

File: lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/ipv-core-main-journey.yaml (hunk 11 above)
Comment (MikeCollingwood) [reply lambdas/process-journey-event/src/main/resources/statemachine/journey-maps/ipv-core-main-journey.yaml:20]: Thank you - I think to avoid potentially breaking a flow, it'd be better done in a dedicated bit of work
